File: quant1x/factors/financial_report.py
# -*- coding: utf-8 -*-
# Copyright (c) Quant1X <wangfengxy@sina.cn>.
# Licensed under the MIT License.

import requests
import pandas as pd
import os
from typing import List, Tuple, Optional, Dict
from dataclasses import dataclass, field
from quant1x.data.market import detect_symbol, correct_security_code, assert_index_by_security_code
from quant1x.data.meta.timestamp import Timestamp
from quant1x import std
from quant1x.config import config, reports_filename

# Constants
URL_QUARTERLY_REPORT_ALL = "https://datacenter-web.eastmoney.com/api/data/v1/get"
EASTMONEY_QUARTERLY_REPORT_ALL_PAGE_SIZE = 50

class FinancialReportException(Exception):
    def __init__(self, code: int, message: str):
        self.code = code
        self.message = message
        super().__init__(f"[{code}] {message}")

@dataclass
class QuarterlyReport:
    SecuCode: str = ""
    UpdateDate: str = ""      # 更新日期
    ReportDate: str = ""      # 报告日期
    NoticeDate: str = ""      # 最新公告日期
    IsNew: str = ""
    ORGCODE: str = ""
    TRADEMARKETZJG: str = ""
    QDATE: str = ""           # 季报期
    DATATYPE: str = ""
    DATAYEAR: str = ""
    DATEMMDD: str = ""
    EITIME: str = ""
    SECURITYCODE: str = ""
    SECURITYNAMEABBR: str = ""
    TRADEMARKETCODE: str = ""
    TRADEMARKET: str = ""
    SECURITYTYPECODE: str = ""
    SECURITYTYPE: str = ""
    BasicEPS: float = 0.0            # 每股收益
    DeductBasicEPS: float = 0.0      # 每股收益(扣除)
    BPS: float = 0.0                 # 每股净资产
    TotalOperateIncome: float = 0.0  # 营业总收入
    ParentNetprofit: float = 0.0     # 净利润
    WeightAvgRoe: float = 0.0        # 净资产收益率
    YSTZ: float = 0.0                # 营业总收入同比增长
    SJLTZ: float = 0.0               # 净利润同比增长
    MGJYXJJE: float = 0.0            # 每股经营现金流量(元)
    XSMLL: float = 0.0               # 销售毛利率(%)
    YSHZ: float = 0.0
    SJLHZ: float = 0.0
    ASSIGNDSCRPT: float = 0.0        # 废弃
    PAYYEAR: float = 0.0             # 废弃
    PUBLISHNAME: str = ""
    ZXGXL: float = 0.0
    SecurityCode: str = ""

@dataclass
class QuarterlyReportSummary:
    QDate: str = ""
    BPS: float = 0.0
    BasicEPS: float = 0.0
    TotalOperateIncome: float = 0.0
    DeductBasicEPS: float = 0.0

    def assign(self, v: QuarterlyReport):
        self.BPS = v.BPS
        self.BasicEPS = v.BasicEPS
        self.TotalOperateIncome = v.TotalOperateIncome
        self.DeductBasicEPS = v.DeductBasicEPS
        self.QDate = v.QDATE

def quarterly_reports(feature_date: str, page_no: int = 1) -> Tuple[List[QuarterlyReport], int, Optional[FinancialReportException]]:
    _, q_begin, q_end = std.get_quarter_by_date(feature_date, 1)
    quarter_end_date = Timestamp.parse(q_end).only_date()

    params = {
        "sortColumns": "REPORTDATE,SECURITY_CODE",
        "sortTypes": "-1,1",
        "pageSize": str(EASTMONEY_QUARTERLY_REPORT_ALL_PAGE_SIZE),
        "pageNumber": str(page_no),
        "reportName": "RPT_LICO_FN_CPD",
        "columns": "ALL",
        "filter": f"(REPORTDATE='{quarter_end_date}')"
    }

    try:
        response = requests.get(URL_QUARTERLY_REPORT_ALL, params=params, timeout=10)
        if response.status_code != 200:
            return [], 0, FinancialReportException(-1, "HTTP请求失败")
        
        raw = response.json()
        result = raw.get("result")
        if not result:
            return [], 0, FinancialReportException(-1, "缺少 result 字段")
            
        pages = result.get("pages", 0)
        data_list = result.get("data", [])
        
        reports = []
        for v in data_list:
            report = QuarterlyReport()
            report.SecuCode = v.get("SECUCODE", "")
            report.UpdateDate = v.get("UPDATE_DATE", "")
            report.ReportDate = v.get("REPORTDATE", "")
            report.NoticeDate = v.get("NOTICE_DATE", "")
            report.IsNew = str(v.get("ISNEW", ""))
            report.ORGCODE = v.get("ORG_CODE", "")
            report.TRADEMARKETZJG = v.get("TRADE_MARKET_ZJG", "")
            report.QDATE = v.get("QDATE", "")
            report.DATATYPE = v.get("DATATYPE", "")
            report.DATAYEAR = v.get("DATAYEAR", "")
            report.DATEMMDD = v.get("DATEMMDD", "")
            report.EITIME = v.get("EITIME", "")
            report.SECURITYCODE = v.get("SECURITY_CODE", "")
            report.SECURITYNAMEABBR = v.get("SECURITY_NAME_ABBR", "")
            report.TRADEMARKETCODE = v.get("TRADE_MARKET_CODE", "")
            report.TRADEMARKET = v.get("TRADE_MARKET", "")
            report.SECURITYTYPECODE = v.get("SECURITY_TYPE_CODE", "")
            report.SECURITYTYPE = v.get("SECURITY_TYPE", "")
            
            report.BasicEPS = float(v.get("BASIC_EPS") or 0.0)
            report.DeductBasicEPS = float(v.get("DEDUCT_BASIC_EPS") or 0.0)
            report.BPS = float(v.get("BPS") or 0.0)
            report.TotalOperateIncome = float(v.get("TOTAL_OPERATE_INCOME") or 0.0)
            report.ParentNetprofit = float(v.get("PARENT_NETPROFIT") or 0.0)
            report.WeightAvgRoe = float(v.get("WEIGHTAVG_ROE") or 0.0)
            report.YSTZ = float(v.get("YSTZ") or 0.0)
            report.SJLTZ = float(v.get("SJLTZ") or 0.0)
            report.MGJYXJJE = float(v.get("MGJYXJJE") or 0.0)
            report.XSMLL = float(v.get("XSMLL") or 0.0)
            report.YSHZ = float(v.get("YSHZ") or 0.0)
            report.SJLHZ = float(v.get("SJLHZ") or 0.0)
            # report.ASSIGNDSCRPT = float(v.get("ASSIGNDSCRPT") or 0.0)
            # report.PAYYEAR = float(v.get("PAYYEAR") or 0.0)
            report.PUBLISHNAME = v.get("PUBLISHNAME", "")
            report.ZXGXL = float(v.get("ZXGXL") or 0.0)
            
            report.SecurityCode = correct_security_code(report.SecuCode)
            reports.append(report)
            
        return reports, pages, None
        
    except Exception as e:
        return [], 0, FinancialReportException(-1, f"JSON解析错误: {str(e)}")

def quarterly_reports_by_security_code(security_code: str, date: str, diff_quarters: int, page_no: int = 1) -> Tuple[List[QuarterlyReport], int, Optional[FinancialReportException]]:
    inst = detect_symbol(security_code)
    code = inst.ticker
    quarter_end_date = Timestamp.parse(date).only_date()
    
    params = {
        "sortColumns": "REPORTDATE,SECURITY_CODE",
        "sortTypes": "-1,1",
        "pageSize": str(EASTMONEY_QUARTERLY_REPORT_ALL_PAGE_SIZE),
        "pageNumber": str(page_no),
        "reportName": "RPT_LICO_FN_CPD",
        "columns": "ALL",
        "filter": f'(SECURITY_CODE="{code}")(REPORTDATE=\'{quarter_end_date}\')'
    }
    
    try:
        response = requests.get(URL_QUARTERLY_REPORT_ALL, params=params, timeout=10)
        if response.status_code != 200:
            return [], 0, FinancialReportException(-1, "HTTP请求失败")
            
        raw = response.json()
        result = raw.get("result")
        if not result:
            return [], 0, FinancialReportException(-1, "缺少 result 字段")
            
        pages = result.get("pages", 0)
        data_list = result.get("data", [])
        
        reports = []
        for v in data_list:
            report = QuarterlyReport()
            report.SecuCode = v.get("SECUCODE", "")
            report.UpdateDate = v.get("UPDATE_DATE", "")
            report.ReportDate = v.get("REPORTDATE", "")
            report.NoticeDate = v.get("NOTICE_DATE", "")
            report.IsNew = str(v.get("ISNEW", ""))
            report.ORGCODE = v.get("ORG_CODE", "")
            report.TRADEMARKETZJG = v.get("TRADE_MARKET_ZJG", "")
            report.QDATE = v.get("QDATE", "")
            report.DATATYPE = v.get("DATATYPE", "")
            report.DATAYEAR = v.get("DATAYEAR", "")
            report.DATEMMDD = v.get("DATEMMDD", "")
            report.EITIME = v.get("EITIME", "")
            report.SECURITYCODE = v.get("SECURITY_CODE", "")
            report.SECURITYNAMEABBR = v.get("SECURITY_NAME_ABBR", "")
            report.TRADEMARKETCODE = v.get("TRADE_MARKET_CODE", "")
            report.TRADEMARKET = v.get("TRADE_MARKET", "")
            report.SECURITYTYPECODE = v.get("SECURITY_TYPE_CODE", "")
            report.SECURITYTYPE = v.get("SECURITY_TYPE", "")
            
            report.BasicEPS = float(v.get("BASIC_EPS") or 0.0)
            report.DeductBasicEPS = float(v.get("DEDUCT_BASIC_EPS") or 0.0)
            report.BPS = float(v.get("BPS") or 0.0)
            report.TotalOperateIncome = float(v.get("TOTAL_OPERATE_INCOME") or 0.0)
            report.ParentNetprofit = float(v.get("PARENT_NETPROFIT") or 0.0)
            report.WeightAvgRoe = float(v.get("WEIGHTAVG_ROE") or 0.0)
            report.YSTZ = float(v.get("YSTZ") or 0.0)
            report.SJLTZ = float(v.get("SJLTZ") or 0.0)
            report.MGJYXJJE = float(v.get("MGJYXJJE") or 0.0)
            report.XSMLL = float(v.get("XSMLL") or 0.0)
            report.YSHZ = float(v.get("YSHZ") or 0.0)
            report.SJLHZ = float(v.get("SJLHZ") or 0.0)
            report.PUBLISHNAME = v.get("PUBLISHNAME", "")
            report.ZXGXL = float(v.get("ZXGXL") or 0.0)
            
            report.SecurityCode = correct_security_code(report.SecuCode)
            reports.append(report)
            
        return reports, pages, None
        
    except Exception as e:
        return [], 0, FinancialReportException(-1, f"JSON解析错误: {str(e)}")

# Cache
_map_reports: Dict[str, List[QuarterlyReport]] = {}

def _quarter_str(date: str) -> str:
    """获取 date 对应的季度字符串，如 2026Q1"""
    quarter, _, _ = std.get_quarter_by_date(date)
    return quarter

def cache_quarterly_reports_by_security_code(date: str, diff_quarters: int = 1) -> Tuple[List[QuarterlyReport], int, Optional[FinancialReportException]]:
    _, _, last = std.get_quarter_by_date(date, diff_quarters)
    filename = reports_filename(last)
    expected_quarter = _quarter_str(last)
    
    # Check memory cache
    if filename in _map_reports:
        return _map_reports[filename], 0, None
        
    # Check file cache
    if os.path.exists(filename):
        try:
            df = pd.read_csv(filename)
            # Validate cache: first row's QDATE must match expected quarter
            if len(df) > 0 and "QDATE" in df.columns:
                cached_qdate = str(df["QDATE"].iloc[0])
                if cached_qdate != expected_quarter:
                    # Stale cache, delete and re-fetch
                    os.remove(filename)
                    df = None
            # Convert DataFrame to List[QuarterlyReport]
            if df is not None:
                reports = []
                for _, row in df.iterrows():
                    r = QuarterlyReport()
                    for field_name in r.__dataclass_fields__:
                        if field_name in row:
                            val = row[field_name]
                            # Handle NaN
                            if pd.isna(val):
                                if isinstance(getattr(r, field_name), (int, float)):
                                    val = 0.0
                                else:
                                    val = ""
                            setattr(r, field_name, val)
                    reports.append(r)
                
                if reports:
                    _map_reports[filename] = reports
                    return reports, 0, None
        except Exception as e:
            print(f"Read CSV error: {e}")
            
    # Fetch from network
    # quarterly_reports internally applies diff_quarters=1 offset,
    # so we pass 'date' directly and use 'last' for cache path/validation.
    qdate = date
    if diff_quarters > 1:
        _, _, tmp_date = std.get_quarter_by_date(date, diff_quarters - 1)
        qdate = tmp_date
        
    all_reports, pages, err = quarterly_reports(qdate)
    if err or pages < 1:
        return [], 0, err
        
    # Fetch remaining pages
    for page_no in range(2, pages + 1):
        tmp_list, tmp_pages, tmp_err = quarterly_reports(qdate, page_no)
        if tmp_err or tmp_pages < 1:
            break
        if not tmp_list:
            break
        all_reports.extend(tmp_list)
        if len(tmp_list) < EASTMONEY_QUARTERLY_REPORT_ALL_PAGE_SIZE:
            break
            
    _map_reports[filename] = all_reports
    
    # Save to CSV
    try:
        os.makedirs(os.path.dirname(filename), exist_ok=True)
        df = pd.DataFrame([vars(r) for r in all_reports])
        df.to_csv(filename, index=False)
    except Exception as e:
        print(f"Save CSV error: {e}")
        
    return all_reports, pages, None

def get_cache_quarterly_reports_by_security_code(security_code: str, date: str, diff_quarters: int = 1) -> Optional[QuarterlyReport]:
    for diff in range(diff_quarters, 5):
        all_reports, _, err = cache_quarterly_reports_by_security_code(date, diff)
        if not err and all_reports:
            for v in all_reports:
                if v.SecurityCode == security_code:
                    return v
    return None

# Individual stock cache
_g_map_quarterly_reports: Dict[str, QuarterlyReport] = {}

def load_quarterly_reports(date: str):
    _, _, last = std.get_quarter_by_date(date, 1)
    filename = reports_filename(last)
    expected_quarter = _quarter_str(last)
    
    if os.path.exists(filename):
        try:
            df = pd.read_csv(filename)
            # Validate cache freshness
            if len(df) > 0 and "QDATE" in df.columns:
                if str(df["QDATE"].iloc[0]) != expected_quarter:
                    os.remove(filename)
                    return
            for _, row in df.iterrows():
                r = QuarterlyReport()
                for field_name in r.__dataclass_fields__:
                    if field_name in row:
                        val = row[field_name]
                        if pd.isna(val):
                            if isinstance(getattr(r, field_name), (int, float)):
                                val = 0.0
                            else:
                                val = ""
                        setattr(r, field_name, val)
                _g_map_quarterly_reports[r.SecurityCode] = r
        except Exception:
            pass

def get_quarterly_report_summary(security_code: str, date: str) -> QuarterlyReportSummary:
    summary = QuarterlyReportSummary()
    
    if assert_index_by_security_code(security_code):
        return summary
        
    if security_code in _g_map_quarterly_reports:
        summary.assign(_g_map_quarterly_reports[security_code])
        return summary
        
    q = get_cache_quarterly_reports_by_security_code(security_code, date)
    if q:
        summary.assign(q)
        
    return summary


if __name__ == "__main__":
    # 测试获取季度财报
    code = "sh600000"
    date = "2026-03-31"
    print(f"=== 测试 quarterly_reports({date}) ===")
    reports, pages, err = quarterly_reports(date, page_no=1)
    if err:
        print(f"  错误: {err.message}")
    else:
        print(f"  共 {len(reports)} 条, {pages} 页")
        for r in reports[:3]:
            print(f"  [{r.SecurityCode}] {r.QDATE} BPS={r.BPS:.2f} EPS={r.BasicEPS:.4f}")

    print(f"\n=== 测试 get_quarterly_report_summary({code}, {date}) ===")
    try:
        summary = get_quarterly_report_summary(code, date)
        print(f"  QDate: {summary.QDate}")
        print(f"  BPS: {summary.BPS}")
        print(f"  BasicEPS: {summary.BasicEPS}")
        print(f"  TotalOperateIncome: {summary.TotalOperateIncome}")
        print(f"  DeductBasicEPS: {summary.DeductBasicEPS}")
    except Exception as e:
        print(f"  错误: {e}")

File: quant1x/factors/financial_report.rs
use serde::{Deserialize, Serialize};
use std::error::Error;
use std::fs;
use std::path::Path;

use crate::config;

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct QuarterlyReport {
    #[serde(rename = "SECURITY_CODE")]
    pub security_code: String,
    #[serde(rename = "UPDATE_DATE")]
    pub update_date: Option<String>,
    #[serde(rename = "REPORTDATE")]
    pub report_date: Option<String>,
    #[serde(rename = "NOTICE_DATE")]
    pub notice_date: Option<String>,
    #[serde(rename = "ISNEW")]
    pub is_new: Option<String>,
    #[serde(rename = "ORG_CODE")]
    pub org_code: Option<String>,
    #[serde(rename = "TRADE_MARKET_ZJG")]
    pub trade_market_zjg: Option<String>,
    #[serde(rename = "QDATE")]
    pub q_date: Option<String>,
    #[serde(rename = "DATATYPE")]
    pub data_type: Option<String>,
    #[serde(rename = "DATAYEAR")]
    pub data_year: Option<String>,
    #[serde(rename = "DATEMMDD")]
    pub date_mmdd: Option<String>,
    #[serde(rename = "EITIME")]
    pub eitime: Option<String>,
    #[serde(rename = "SECUCODE")]
    pub secu_code: Option<String>,
    #[serde(rename = "SECURITY_NAME_ABBR")]
    pub security_name_abbr: Option<String>,
    #[serde(rename = "TRADE_MARKET_CODE")]
    pub trade_market_code: Option<String>,
    #[serde(rename = "TRADE_MARKET")]
    pub trade_market: Option<String>,
    #[serde(rename = "SECURITY_TYPE_CODE")]
    pub security_type_code: Option<String>,
    #[serde(rename = "SECURITY_TYPE")]
    pub security_type: Option<String>,
    #[serde(rename = "BASIC_EPS")]
    pub basic_eps: Option<f64>,
    #[serde(rename = "DEDUCT_BASIC_EPS")]
    pub deduct_basic_eps: Option<f64>,
    #[serde(rename = "BPS")]
    pub bps: Option<f64>,
    #[serde(rename = "TOTAL_OPERATE_INCOME")]
    pub total_operate_income: Option<f64>,
    #[serde(rename = "PARENT_NETPROFIT")]
    pub parent_netprofit: Option<f64>,
    #[serde(rename = "WEIGHTAVG_ROE")]
    pub weight_avg_roe: Option<f64>,
    #[serde(rename = "YSTZ")]
    pub ystz: Option<f64>,
    #[serde(rename = "SJLTZ")]
    pub sjltz: Option<f64>,
    #[serde(rename = "MGJYXJJE")]
    pub mgjyxjje: Option<f64>,
    #[serde(rename = "XSMLL")]
    pub xsmll: Option<f64>,
    #[serde(rename = "YSHZ")]
    pub yshz: Option<f64>,
    #[serde(rename = "SJLHZ")]
    pub sjlhz: Option<f64>,
    #[serde(rename = "ASSIGNDSCRPT")]
    pub assign_dscrpt: Option<String>,
    #[serde(rename = "PAYYEAR")]
    pub pay_year: Option<String>,
    #[serde(rename = "PUBLISHNAME")]
    pub publish_name: Option<String>,
    #[serde(rename = "ZXGXL")]
    pub zxgxl: Option<f64>,
    #[serde(rename = "BOARD_NAME")]
    pub board_name: Option<String>,
    #[serde(rename = "ORI_BOARD_CODE")]
    pub ori_board_code: Option<String>,
    #[serde(rename = "BOARD_CODE")]
    pub board_code: Option<String>,
}

#[derive(Debug, Deserialize)]
struct ApiResponse {
    version: Option<String>,
    result: Option<ApiResult>,
    success: Option<bool>,
    message: Option<String>,
    code: Option<i32>,
}

#[derive(Debug, Deserialize)]
struct ApiResult {
    pages: Option<i32>,
    data: Option<Vec<QuarterlyReport>>,
    count: Option<i32>,
}

const URL_QUARTERLY_REPORT_ALL: &str = "https://datacenter-web.eastmoney.com/api/data/v1/get";
const PAGE_SIZE: i32 = 50;

pub fn fetch_quarterly_reports(
    feature_date: &str,
    page_no: i32,
) -> Result<(Vec<QuarterlyReport>, i32), Box<dyn Error>> {
    let filter = format!("(REPORTDATE='{}')", feature_date);
    let params = [
        ("sortColumns", "REPORTDATE"),
        ("sortTypes", "-1"),
        ("pageSize", &PAGE_SIZE.to_string()),
        ("pageNumber", &page_no.to_string()),
        ("columns", "ALL"),
        ("filter", &filter),
        ("reportName", "RPT_LICO_FN_CPD"),
    ];

    let client = reqwest::blocking::Client::new();
    let resp = client.get(URL_QUARTERLY_REPORT_ALL).query(&params).send()?;

    if !resp.status().is_success() {
        return Err(format!("HTTP error: {}", resp.status()).into());
    }

    let text = resp.text()?;
    let api_resp: ApiResponse = serde_json::from_str(&text)?;

    if let Some(result) = api_resp.result {
        let data = result.data.unwrap_or_default();
        let pages = result.pages.unwrap_or(0);
        Ok((data, pages))
    } else {
        Ok((Vec::new(), 0))
    }
}

pub fn fetch_quarterly_reports_by_security_code(
    security_code: &str,
    page_no: i32,
) -> Result<(Vec<QuarterlyReport>, i32), Box<dyn Error>> {
    let filter = format!("(SECURITY_CODE=\"{}\")", security_code);
    let params = [
        ("sortColumns", "REPORTDATE"),
        ("sortTypes", "-1"),
        ("pageSize", &PAGE_SIZE.to_string()),
        ("pageNumber", &page_no.to_string()),
        ("columns", "ALL"),
        ("filter", &filter),
        ("reportName", "RPT_LICO_FN_CPD"),
    ];

    let client = reqwest::blocking::Client::new();
    let resp = client.get(URL_QUARTERLY_REPORT_ALL).query(&params).send()?;

    if !resp.status().is_success() {
        return Err(format!("HTTP error: {}", resp.status()).into());
    }

    let text = resp.text()?;
    let api_resp: ApiResponse = serde_json::from_str(&text)?;

    if let Some(result) = api_resp.result {
        let data = result.data.unwrap_or_default();
        let pages = result.pages.unwrap_or(0);
        Ok((data, pages))
    } else {
        Ok((Vec::new(), 0))
    }
}

pub fn load_quarterly_reports(date: &str) -> Result<(), Box<dyn Error>> {
    let filename = config::reports_filename(date);
    let path = Path::new(&filename);

    if path.exists() {
        // Already cached
        return Ok(());
    }

    // Ensure directory exists
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)?;
    }

    let mut all_reports = Vec::new();
    let mut page = 1;
    loop {
        let (reports, total_pages) = fetch_quarterly_reports(date, page)?;
        if reports.is_empty() {
            break;
        }
        all_reports.extend(reports);
        if page >= total_pages {
            break;
        }
        page += 1;
    }

    // Save to CSV
    let mut wtr = csv::Writer::from_path(path)?;
    for report in all_reports {
        wtr.serialize(report)?;
    }
    wtr.flush()?;

    Ok(())
}

#[derive(Debug, Clone, Default)]
pub struct QuarterlyReportSummary {
    pub q_date: String,
    pub bps: f64,
    pub basic_eps: f64,
    pub total_operate_income: f64,
    pub deduct_basic_eps: f64,
}

impl QuarterlyReportSummary {
    pub fn from_report(report: &QuarterlyReport) -> Self {
        Self {
            q_date: report.q_date.clone().unwrap_or_default(),
            bps: report.bps.unwrap_or(0.0),
            basic_eps: report.basic_eps.unwrap_or(0.0),
            total_operate_income: report.total_operate_income.unwrap_or(0.0),
            deduct_basic_eps: report.deduct_basic_eps.unwrap_or(0.0),
        }
    }
}

pub fn get_quarterly_report_summary(
    security_code: &str,
    date: &str,
) -> Option<QuarterlyReportSummary> {
    let filename = config::reports_filename(date);
    let path = Path::new(&filename);

    if !path.exists() {
        // Try to load if not exists
        if let Err(e) = load_quarterly_reports(date) {
            log::error!("Failed to load quarterly reports for {}: {}", date, e);
            return None;
        }
    }

    let mut rdr = match csv::Reader::from_path(path) {
        Ok(r) => r,
        Err(e) => {
            log::error!("Failed to read CSV {}: {}", filename, e);
            return None;
        }
    };

    for result in rdr.deserialize() {
        let report: QuarterlyReport = match result {
            Ok(r) => r,
            Err(_) => continue,
        };
        if report.security_code == security_code {
            return Some(QuarterlyReportSummary::from_report(&report));
        }
    }

    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_quarterly_report_struct() {
        let json = r#"{
            "SECURITY_CODE": "301381",
            "SECURITY_NAME_ABBR": "赛维时代",
            "TRADE_MARKET_CODE": "069001002002",
            "TRADE_MARKET": "深交所创业板",
            "SECURITY_TYPE_CODE": "058001001",
            "SECURITY_TYPE": "A股",
            "UPDATE_DATE": "2025-04-26 00:00:00",
            "REPORTDATE": "2025-03-31 00:00:00",
            "BASIC_EPS": 0.1175,
            "TOTAL_OPERATE_INCOME": 2458280774.74
        }"#;

        let report: QuarterlyReport = serde_json::from_str(json).unwrap();
        assert_eq!(report.security_code, "301381");
        assert_eq!(report.basic_eps, Some(0.1175));
        assert_eq!(report.total_operate_income, Some(2458280774.74));
    }
}
